### PR TITLE
[SDA] allow asset namespace to be a list

### PR DESCRIPTION
**Pull request by OwenKephart**
Without this change, asset keys generated using @asset can only have two "segments" at most (namespace, name of function).

We probably want to be more flexible than that.

This preserves the old behavior when users input a single string (so it's backwards compatible).

### CODE REVIEW

File: python_modules/dagster/dagster/core/asset_defs/decorators.py [hunk 1]
```diff
@@ -295,8 +296,8 @@ def build_asset_ins(
         )
 
     for asset_key in non_argument_deps:
-        stringified_asset_key = asset_key.to_string(legacy=True)
+        stringified_asset_key = "_".join(asset_key.path)
```
Comment (clairelin135) [inline python_modules/dagster/dagster/core/asset_defs/decorators.py:299]: Why the change to this join? The param name doesn't need to exist in the asset definition signature, so it might be nice to use the `to_string` for consistency

**Review by clairelin135 — COMMENTED** [review]
Thanks for making this change, just had one question (inline)

File: python_modules/dagster/dagster/core/asset_defs/decorators.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/core/asset_defs/decorators.py:299]: asset_key.to_string() returns (basically) `".".join(asset_key.path)`. so it actually will raise an error if you have an asset key with multiple segments to the key (i.e. `AssetKey(["foo", "bar"])`), because "foo.bar" is not a valid identifier (but "foo_bar" is).

Comment (clairelin135) [reply python_modules/dagster/dagster/core/asset_defs/decorators.py:299]: Oh, I see. thanks for fixing!

**Review by clairelin135 — APPROVED** [review]
